File: Free SysLog/Support Code/Namespace Code/SaveAppSettings.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text.RegularExpressions;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace Free_SysLog.SaveAppSettings
{
    public static class SaveAppSettings
    {
        public static void SaveApplicationSettingsToFile(string strFileName)
        {
            var exportedSettingsArray = new Dictionary<string, object>((int)StringComparison.OrdinalIgnoreCase);
            Type settingType;
            Point point;
            Size size;
            object rawValue;

            foreach (System.Configuration.SettingsPropertyValue settingProperty in My.MySettingsProperty.Settings.PropertyValues)
            {
                if (settingProperty.PropertyValue is not null)
                {
                    settingType = settingProperty.PropertyValue.GetType();

                    if (settingType == typeof(Point))
                    {
                        point = (Point)settingProperty.PropertyValue;
                        rawValue = $"{point.X}|{point.Y}";
                        point = default;
                    }
                    else if (settingType == typeof(Color))
                    {
                        rawValue = ((Color)settingProperty.PropertyValue).ToArgb();
                    }
                    else if (settingType == typeof(Size))
                    {
                        size = (Size)settingProperty.PropertyValue;
                        rawValue = $"{size.Height}|{size.Width}";
                        size = default;
                    }
                    else
                    {
                        rawValue = settingProperty.PropertyValue;
                    }

                    exportedSettingsArray.Add(settingProperty.Name.Trim(), rawValue);
                }
            }

            using (var streamWriter = new System.IO.StreamWriter(strFileName))
            {
                streamWriter.Write(Newtonsoft.Json.JsonConvert.SerializeObject(exportedSettingsArray, Newtonsoft.Json.Formatting.Indented));
            }
        }

        private static Font ParseFontFromString(string fontString)
        {
            try
            {
                var MatchResults = Regex.Match(fontString, @"(?<fontname>[^\n\r,]+), (?<size>[0-9]+\.[0-9]{2})[pt]{2}(?:, style=)?(?<style>.*)", RegexOptions.IgnoreCase);

                if (MatchResults.Success)
                {
                    string fontName = MatchResults.Groups["fontname"].Value;
                    float fontSize = 8.25f;
                    var fontStyle = FontStyle.Regular;

                    if (!float.TryParse(MatchResults.Groups["size"].Value, out fontSize))
                        fontSize = 8.25f;

                    if (!string.IsNullOrWhiteSpace(MatchResults.Groups["style"].Value))
                    {
                        string strStyleValue = MatchResults.Groups["style"].Value;

                        if (strStyleValue.CaseInsensitiveContains("Bold"))
                            fontStyle = fontStyle | FontStyle.Bold;
                        if (strStyleValue.CaseInsensitiveContains("Italic"))
                            fontStyle = fontStyle | FontStyle.Italic;
                        if (strStyleValue.CaseInsensitiveContains("Underline"))
                            fontStyle = fontStyle | FontStyle.Underline;
                        if (strStyleValue.CaseInsensitiveContains("Strikeout"))
                            fontStyle = fontStyle | FontStyle.Strikeout;
                    }

                    return new Font(fontName, fontSize, fontStyle);
                }
                else
                {
                    return new Font("Microsoft Sans Serif", 9.75f);
                }
            }
            catch (Exception ex)
            {
                return new Font("Microsoft Sans Serif", 9.75f);
            }
        }

        public static bool LoadApplicationSettingsFromFile(string strFileName, string strMessageBoxTitle)
        {
            try
            {
                var exportedSettingsArray = new Dictionary<string, object>((int)StringComparison.OrdinalIgnoreCase);
                bool boolResult;
                byte byteResult;
                int intResult;
                long longResult;
                Type settingType;
                short shortResult;
                string[] splitArray;
                object rawValue = null;

                using (var streamReader = new System.IO.StreamReader(strFileName))
                {
                    exportedSettingsArray = Newtonsoft.Json.JsonConvert.DeserializeObject<Dictionary<string, object>>(streamReader.ReadToEnd().Trim(), SupportCode.SupportCode.JSONDecoderSettingsForSettingsFiles);
                }

                My.MySettingsProperty.Settings.replacements = new System.Collections.Specialized.StringCollection();
                My.MySettingsProperty.Settings.ServersToSendTo = new System.Collections.Specialized.StringCollection();
                My.MySettingsProperty.Settings.alerts = new System.Collections.Specialized.StringCollection();
                My.MySettingsProperty.Settings.ignored2 = new System.Collections.Specialized.StringCollection();
                My.MySettingsProperty.Settings.hostnames = new System.Collections.Specialized.StringCollection();

                foreach (System.Configuration.SettingsPropertyValue settingProperty in My.MySettingsProperty.Settings.PropertyValues)
                {
                    if (exportedSettingsArray.FindKeyInDictionaryAndReturnIt(settingProperty.Name, ref rawValue) & settingProperty.PropertyValue is not null)
                    {
                        if (settingProperty.Name.Equals("listFilesColumnOrder", StringComparison.OrdinalIgnoreCase) | settingProperty.Name.Equals("verifyListFilesColumnOrder", StringComparison.OrdinalIgnoreCase))
                        {
                            settingType = typeof(System.Collections.Specialized.StringCollection);
                        }
                        else
                        {
                            settingType = settingProperty.PropertyValue.GetType();
                        }

                        if (settingType == typeof(Color) && int.TryParse(Conversions.ToString(rawValue), out intResult))
                        {
                            My.MySettingsProperty.Settings[settingProperty.Name] = Color.FromArgb(intResult);
                        }
                        else if (settingType == typeof(Point))
                        {
                            splitArray = rawValue.ToString().Split('|');
                            My.MySettingsProperty.Settings[settingProperty.Name] = new Point() { X = Conversions.ToInteger(splitArray[0]), Y = Conversions.ToInteger(splitArray[1]) };
                            splitArray = null;
                        }
                        else if (settingType == typeof(Font))
                        {
                            My.MySettingsProperty.Settings[settingProperty.Name] = ParseFontFromString(Conversions.ToString(rawValue));
                        }
                        else if (settingType == typeof(Size))
                        {
                            splitArray = rawValue.ToString().Split('|');
                            My.MySettingsProperty.Settings[settingProperty.Name] = new Size() { Height = Conversions.ToInteger(splitArray[0]), Width = Conversions.ToInteger(splitArray[1]) };
                            splitArray = null;
                        }
                        else if (settingType == typeof(bool) && bool.TryParse(Conversions.ToString(rawValue), out boolResult))
                        {
                            My.MySettingsProperty.Settings[settingProperty.Name] = boolResult;
                        }
                        else if (settingType == typeof(byte) && byte.TryParse(Conversions.ToString(rawValue), out byteResult))
                        {
                            My.MySettingsProperty.Settings[settingProperty.Name] = byteResult;
                        }
                        else if (settingType == typeof(short) && short.TryParse(Conversions.ToString(rawValue), out shortResult))
                        {
                            My.MySettingsProperty.Settings[settingProperty.Name] = shortResult;
                        }
                        else if (settingType == typeof(int) && int.TryParse(Conversions.ToString(rawValue), out intResult))
                        {
                            My.MySettingsProperty.Settings[settingProperty.Name] = intResult;
                        }
                        else if (settingType == typeof(long) && long.TryParse(Conversions.ToString(rawValue), out longResult))
                        {
                            My.MySettingsProperty.Settings[settingProperty.Name] = longResult;
                        }
                        else if (settingType == typeof(System.Collections.Specialized.StringCollection))
                        {
                            My.MySettingsProperty.Settings[settingProperty.Name] = ConvertArrayListToSpecializedStringCollection((Newtonsoft.Json.Linq.JArray)rawValue);
                        }
                        else
                        {
                            My.MySettingsProperty.Settings[settingProperty.Name] = rawValue;
                        }
                    }
                }

                return true;
            }
            catch (Exception ex)
            {
                Interaction.MsgBox($"There was an issue decoding your chosen JSON settings file, import failed.{Constants.vbCrLf}{Constants.vbCrLf}{ex.Message}{ex.StackTrace.Trim()}", MsgBoxStyle.Critical, strMessageBoxTitle);
                return false;
            }
        }

        private static System.Collections.Specialized.StringCollection ConvertArrayListToSpecializedStringCollection(Newtonsoft.Json.Linq.JArray input)
        {
            try
            {
                var stringCollection = new System.Collections.Specialized.StringCollection();
                foreach (string item in input)
                    stringCollection.Add(item);
                return stringCollection;
            }
            catch (Exception ex)
            {
                return new System.Collections.Specialized.StringCollection();
            }
        }

        /// <summary>This function operates a lot like ContainsKey() but is case-InSeNsItIvE and it returns the value of the key/value pair if the function returns True.</summary>
        /// <param name="haystack">The dictionary that's being searched.</param>
        /// <param name="needle">The key that you're looking for.</param>
        /// <param name="value">The value of the key you're looking for, passed as a ByRef so that you can access the value if the function returns True.</param>
        /// <return>Returns a String value.</return>
        private static bool FindKeyInDictionaryAndReturnIt(this Dictionary<string, object> haystack, string needle, ref object value)
        {
            if (string.IsNullOrEmpty(needle))
            {
                throw new ArgumentException($"'{nameof(needle)}' cannot be null or empty.", nameof(needle));
            }
            if (haystack is null)
            {
                throw new ArgumentNullException(nameof(haystack));
            }

            var KeyValuePair = haystack.FirstOrDefault((item) => item.Key.Trim().Equals(needle, StringComparison.OrdinalIgnoreCase));
            if (KeyValuePair.Value is not null)
                value = KeyValuePair.Value;
            return KeyValuePair.Value is not null;
        }
    }
}

File: Free SysLog/Support Code/Namespace Code/SaveAppSettings.vb
Imports System.Runtime.CompilerServices
Imports System.Text.RegularExpressions
Imports Free_SysLog.SupportCode

Namespace SaveAppSettings
    Public Module SaveAppSettings
        Public Sub SaveApplicationSettingsToFile(strFileName As String)
            Dim exportedSettingsArray As New Dictionary(Of String, Object)(StringComparison.OrdinalIgnoreCase)
            Dim settingType As Type
            Dim point As Point, size As Size
            Dim rawValue As Object

            For Each settingProperty As Configuration.SettingsPropertyValue In My.Settings.PropertyValues
                If settingProperty.PropertyValue IsNot Nothing Then
                    settingType = settingProperty.PropertyValue.GetType

                    If settingType = GetType(Point) Then
                        point = DirectCast(settingProperty.PropertyValue, Point)
                        rawValue = $"{point.X}|{point.Y}"
                        point = Nothing
                    ElseIf settingType = GetType(Color) Then
                        rawValue = DirectCast(settingProperty.PropertyValue, Color).ToArgb
                    ElseIf settingType = GetType(Size) Then
                        size = DirectCast(settingProperty.PropertyValue, Size)
                        rawValue = $"{size.Height}|{size.Width}"
                        size = Nothing
                    Else
                        rawValue = settingProperty.PropertyValue
                    End If

                    exportedSettingsArray.Add(settingProperty.Name.Trim, rawValue)
                End If
            Next

            Using streamWriter As New IO.StreamWriter(strFileName)
                streamWriter.Write(Newtonsoft.Json.JsonConvert.SerializeObject(exportedSettingsArray, Newtonsoft.Json.Formatting.Indented))
            End Using
        End Sub

        Private Function ParseFontFromString(fontString As String) As Font
            Try
                Dim MatchResults As Match = Regex.Match(fontString, "(?<fontname>[^\n\r,]+), (?<size>[0-9]+\.[0-9]{2})[pt]{2}(?:, style=)?(?<style>.*)", RegexOptions.IgnoreCase)

                If MatchResults.Success Then
                    Dim fontName As String = MatchResults.Groups("fontname").Value
                    Dim fontSize As Single = 8.25
                    Dim fontStyle As FontStyle = FontStyle.Regular

                    If Not Single.TryParse(MatchResults.Groups("size").Value, fontSize) Then fontSize = 8.25

                    fontSize = Math.Max(6.0F, Math.Min(fontSize, 32.0F))

                    If Not String.IsNullOrWhiteSpace(MatchResults.Groups("style").Value) Then
                        Dim strStyleValue As String = MatchResults.Groups("style").Value

                        If strStyleValue.CaseInsensitiveContains("Bold") Then fontStyle = fontStyle Or FontStyle.Bold
                        If strStyleValue.CaseInsensitiveContains("Italic") Then fontStyle = fontStyle Or FontStyle.Italic
                        If strStyleValue.CaseInsensitiveContains("Underline") Then fontStyle = fontStyle Or FontStyle.Underline
                        If strStyleValue.CaseInsensitiveContains("Strikeout") Then fontStyle = fontStyle Or FontStyle.Strikeout
                    End If

                    Return New Font(fontName, fontSize, fontStyle)
                Else
                    Return Control.DefaultFont
                End If
            Catch ex As Exception
                Return Control.DefaultFont
            End Try
        End Function

        Public Function LoadApplicationSettingsFromFile(strFileName As String, strMessageBoxTitle As String) As Boolean
            Try
                Dim exportedSettingsArray As New Dictionary(Of String, Object)(StringComparison.OrdinalIgnoreCase)
                Dim boolResult As Boolean, byteResult As Byte, intResult As Integer, longResult As Long, settingType As Type, shortResult As Short, splitArray As String()
                Dim rawValue As Object = Nothing

                Using streamReader As New IO.StreamReader(strFileName)
                    exportedSettingsArray = Newtonsoft.Json.JsonConvert.DeserializeObject(Of Dictionary(Of String, Object))(streamReader.ReadToEnd.Trim, JSONDecoderSettingsForSettingsFiles)
                End Using

                My.Settings.replacements = New Specialized.StringCollection()
                My.Settings.ServersToSendTo = New Specialized.StringCollection()
                My.Settings.alerts = New Specialized.StringCollection()
                My.Settings.ignored2 = New Specialized.StringCollection()
                My.Settings.hostnames = New Specialized.StringCollection()

                For Each settingProperty As Configuration.SettingsPropertyValue In My.Settings.PropertyValues
                    If exportedSettingsArray.FindKeyInDictionaryAndReturnIt(settingProperty.Name, rawValue) And settingProperty.PropertyValue IsNot Nothing Then
                        If settingProperty.Name.Equals("listFilesColumnOrder", StringComparison.OrdinalIgnoreCase) Or settingProperty.Name.Equals("verifyListFilesColumnOrder", StringComparison.OrdinalIgnoreCase) Then
                            settingType = GetType(Specialized.StringCollection)
                        Else
                            settingType = settingProperty.PropertyValue.GetType
                        End If

                        If settingType = GetType(Color) AndAlso Integer.TryParse(rawValue, intResult) Then
                            My.Settings(settingProperty.Name) = Color.FromArgb(intResult)
                        ElseIf settingType = GetType(Point) Then
                            splitArray = rawValue.split("|")
                            My.Settings(settingProperty.Name) = New Point() With {.X = splitArray(0), .Y = splitArray(1)}
                            splitArray = Nothing
                        ElseIf settingType = GetType(Font) Then
                            My.Settings(settingProperty.Name) = ParseFontFromString(rawValue)
                        ElseIf settingType = GetType(Size) Then
                            splitArray = rawValue.split("|")
                            My.Settings(settingProperty.Name) = New Size() With {.Height = splitArray(0), .Width = splitArray(1)}
                            splitArray = Nothing
                        ElseIf settingType = GetType(Boolean) AndAlso Boolean.TryParse(rawValue, boolResult) Then
                            My.Settings(settingProperty.Name) = boolResult
                        ElseIf settingType = GetType(Byte) AndAlso Byte.TryParse(rawValue, byteResult) Then
                            My.Settings(settingProperty.Name) = byteResult
                        ElseIf settingType = GetType(Short) AndAlso Short.TryParse(rawValue, shortResult) Then
                            My.Settings(settingProperty.Name) = shortResult
                        ElseIf settingType = GetType(Integer) AndAlso Integer.TryParse(rawValue, intResult) Then
                            My.Settings(settingProperty.Name) = intResult
                        ElseIf settingType = GetType(Long) AndAlso Long.TryParse(rawValue, longResult) Then
                            My.Settings(settingProperty.Name) = longResult
                        ElseIf settingType = GetType(Specialized.StringCollection) Then
                            My.Settings(settingProperty.Name) = ConvertArrayListToSpecializedStringCollection(rawValue)
                        Else
                            My.Settings(settingProperty.Name) = rawValue
                        End If
                    End If
                Next

                Return True
            Catch ex As Exception
                MsgBox($"There was an issue decoding your chosen JSON settings file, import failed.{vbCrLf}{vbCrLf}{ex.Message}{ex.StackTrace.Trim}", MsgBoxStyle.Critical, strMessageBoxTitle)
                Return False
            End Try
        End Function

        Private Function ConvertArrayListToSpecializedStringCollection(input As Newtonsoft.Json.Linq.JArray) As Specialized.StringCollection
            If input Is Nothing Then Return New Specialized.StringCollection

            Dim item As String
            Dim stringCollection As New Specialized.StringCollection

            For Each token As Newtonsoft.Json.Linq.JToken In input
                item = token?.ToString()

                If Not String.IsNullOrEmpty(item) Then
                    Try
                        Newtonsoft.Json.JsonConvert.DeserializeObject(item) ' Validate the JSON.
                        stringCollection.Add(item) ' Good, it passed the decoding so we add it to the collection.
                    Catch ex As Newtonsoft.Json.JsonException
                        ' Something happened while validating the JSON, so we skip the entry.
                    End Try
                End If
            Next

            Return stringCollection
        End Function

        ''' <summary>This function operates a lot like ContainsKey() but is case-InSeNsItIvE and it returns the value of the key/value pair if the function returns True.</summary>
        ''' <param name="haystack">The dictionary that's being searched.</param>
        ''' <param name="needle">The key that you're looking for.</param>
        ''' <param name="value">The value of the key you're looking for, passed as a ByRef so that you can access the value if the function returns True.</param>
        ''' <return>Returns a String value.</return>
        <Extension()>
        Private Function FindKeyInDictionaryAndReturnIt(haystack As Dictionary(Of String, Object), needle As String, ByRef value As Object) As Boolean
            If String.IsNullOrEmpty(needle) Then
                Throw New ArgumentException($"'{NameOf(needle)}' cannot be null or empty.", NameOf(needle))
            End If
            If haystack Is Nothing Then
                Throw New ArgumentNullException(NameOf(haystack))
            End If

            Dim KeyValuePair As KeyValuePair(Of String, Object) = haystack.FirstOrDefault(Function(item As KeyValuePair(Of String, Object)) item.Key.Trim.Equals(needle, StringComparison.OrdinalIgnoreCase))
            If KeyValuePair.Value IsNot Nothing Then value = KeyValuePair.Value
            Return KeyValuePair.Value IsNot Nothing
        End Function
    End Module
End Namespace